pkg/controller/apiextensions/composite/api.go

```diff
@@ -125,6 +155,68 @@ func (r *APISelectorResolver) ResolveSelector(ctx context.Context, cp resource.C
 	return errors.New(errNoCompatibleComposition)
 }
 
+// NewAPIDefaultCompositionSelector returns a APIDefaultCompositionSelector.
+func NewAPIDefaultCompositionSelector(c client.Client, ref v1.ObjectReference, r event.Recorder) *APIDefaultCompositionSelector {
+	return &APIDefaultCompositionSelector{client: c, defRef: ref, recorder: r}
+}
+
+// APIDefaultCompositionSelector selects the default composition referenced in
+// the definition of the resource if neither a reference nor selector is given
+// in composite resource.
+type APIDefaultCompositionSelector struct {
+	client   client.Client
+	defRef   v1.ObjectReference
+	recorder event.Recorder
+}
+
+// SelectComposition selects the default compositionif neither a reference nor
+// selector is given in composite resource.
+func (s *APIDefaultCompositionSelector) SelectComposition(ctx context.Context, cp resource.Composite) error {
+	if cp.GetCompositionReference() != nil || cp.GetCompositionSelector() != nil {
+		return nil
+	}
+	def := &v1alpha1.InfrastructureDefinition{}
+	if err := s.client.Get(ctx, meta.NamespacedNameOf(&s.defRef), def); err != nil {
+		return errors.Wrap(err, errGetInfraDef)
+	}
+	if def.Spec.DefaultCompositionRef == nil {
+		return nil
+	}
+	cp.SetCompositionReference(def.Spec.DefaultCompositionRef)
+	s.recorder.Event(cp, event.Normal(reasonCompositionSelection, "Default composition has been selected"))
```

Comment:
Is this information really worth plumbing a recorder down here for? Would we still be able to record "A composition has been selected" without specifically communicating that it was a default or enforced composition if we did this further up in the stack?

Reply:
I think what composition is selected is already clear from its field on `spec` but I find it valuable to see how it's selected to avoid situations where it's relatively hard to figure out what's happening. This could happen in this situation because the person who is creating a requirement or composite resource is kind of far from seeing what default/enforced composition is easily, it's like you have to check the CRD to see it which we never really do. It's also good to educate people who are exploring how stuff works at the place where they do that exploration directly. So, I found it helpful to say how it made it here, if it's exceptional like default/enforced, and what is selected is already shown in spec when you do `kubectl describe`